File: GO/Structures/QueueLinear.go
package Structures

type QueueLinear struct {
    data     []int
    front    int
    rear     int
    size     int
    capacity int
}

func NewQueueLinear(capacity int) *QueueLinear {
    return &QueueLinear{
        data:     make([]int, capacity),
        front:    0,
        rear:     -1,
        size:     0,
        capacity: capacity,
    }
}

func (q *QueueLinear) Enqueue(value int) {
    if q.size < q.capacity {
        q.rear = (q.rear + 1) % q.capacity
        q.data[q.rear] = value
        q.size++
    }
}

func (q *QueueLinear) Dequeue() (int, bool) {
    if q.size == 0 {
        return 0, false
    }
    
    value := q.data[q.front]
    q.front = (q.front + 1) % q.capacity
    q.size--
    return value, true
}

func (q *QueueLinear) ToArray() []int {
    result := make([]int, q.size)
    for i := 0; i < q.size; i++ {
        index := (q.front + i) % q.capacity
        result[i] = q.data[index]
    }
    return result
}

func (q *QueueLinear) FromArray(arr []int) {
    q.front = 0
    q.rear = -1
    q.size = 0
    
    for _, v := range arr {
        q.Enqueue(v)
    }
}

func (q *QueueLinear) Size() int {
    return q.size
}

File: PYTHON/structures/queue_linear.py
# structures/queue_linear.py
from typing import List, Tuple
import sys
sys.path.append('..')
from utils.data_structure import DataStructure


class QueueLinear(DataStructure):
    """Fila com alocação linear (array circular)"""
    
    def __init__(self, capacity: int):
        self.data = [0] * capacity
        self.front = 0
        self.rear = -1
        self._size = 0
        self.capacity = capacity
    
    def enqueue(self, value: int) -> None:
        if self._size < self.capacity:
            self.rear = (self.rear + 1) % self.capacity
            self.data[self.rear] = value
            self._size += 1
    
    def dequeue(self) -> Tuple[int, bool]:
        if self._size == 0:
            return 0, False
        
        value = self.data[self.front]
        self.front = (self.front + 1) % self.capacity
        self._size -= 1
        return value, True
    
    def to_array(self) -> List[int]:
        result = []
        for i in range(self._size):
            index = (self.front + i) % self.capacity
            result.append(self.data[index])
        return result
    
    def from_array(self, arr: List[int]) -> None:
        self.front = 0
        self.rear = -1
        self._size = 0
        
        for v in arr:
            self.enqueue(v)
    
    def size(self) -> int:
        return self._size